File: js/producer.js
import express from "express";
import amqp from "amqplib";
import winston from "winston";

// Configure logging
const logger = winston.createLogger({
  level: "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.json(),
  ),
  transports: [new winston.transports.Console()],
});

const QUEUE_NAME = "Server1Queue";
const RABBITMQ_URL = "amqp://guest:guest@localhost:5672";

class RabbitMQProducer {
  constructor() {
    this.connection = null;
    this.channel = null;
  }

  async connect() {
    try {
      this.connection = await amqp.connect(RABBITMQ_URL);
      this.channel = await this.connection.createChannel();

      await this.channel.assertQueue(QUEUE_NAME, {
        durable: true,
      });

      logger.info("RabbitMQ connection established");
    } catch (err) {
      logger.error(`RabbitMQ connection failed: ${err}`);
      throw err;
    }
  }

  async publishMessage(message) {
    try {
      const sent = this.channel.sendToQueue(QUEUE_NAME, Buffer.from(message), {
        persistent: true,
      });

      if (!sent) {
        throw new Error("Message rejected by RabbitMQ");
      }

      logger.info(`Message published: ${message}`);
      return true;
    } catch (err) {
      logger.error(`Message publish failed: ${err}`);
      throw err;
    }
  }

  async close() {
    try {
      if (this.channel) await this.channel.close();
      if (this.connection) await this.connection.close();
      logger.info("RabbitMQ connection closed");
    } catch (err) {
      logger.error(`Error closing connection: ${err}`);
    }
  }
}

// Express App Setup
const app = express();
const producer = new RabbitMQProducer();

// Health check endpoint
app.get("/health", (req, res) => {
  res.status(200).json({ status: "healthy" });
});

// Message endpoint
app.get("/send", async (req, res) => {
  const { msg } = req.query;

  if (!msg) {
    logger.warn("Message validation failed");
    return res.status(400).json({ error: "Message is required" });
  }

  try {
    await producer.publishMessage(msg);
    res.status(200).json({ message: msg, status: "success" });
  } catch (err) {
    logger.error(`API error: ${err}`);
    res.status(500).json({ error: "Failed to publish message" });
  }
});

// Startup and Shutdown
async function startServer() {
  try {
    await producer.connect();

    const server = app.listen(8080, () => {
      logger.info("Producer service running on port 8080");
    });

    // Graceful shutdown
    process.on("SIGTERM", async () => {
      logger.info("SIGTERM received - shutting down");
      await producer.close();
      server.close(() => {
        process.exit(0);
      });
    });

    process.on("SIGINT", async () => {
      logger.info("SIGINT received - shutting down");
      await producer.close();
      server.close(() => {
        process.exit(0);
      });
    });
  } catch (err) {
    logger.error(`Startup failed: ${err}`);
    process.exit(1);
  }
}

startServer();

File: py/producer.py
import logging
from contextlib import asynccontextmanager
from typing import Optional

import pika
from fastapi import FastAPI, HTTPException, status
from fastapi.responses import JSONResponse
from pika.adapters.blocking_connection import BlockingChannel

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

QUEUE_NAME = "Server1Queue"
RABBITMQ_URL = "amqp://guest:guest@localhost:5672"


def setup_rabbitmq() -> BlockingChannel:
    """Establish RabbitMQ Connection and channel"""
    try:
        connection = pika.BlockingConnection(pika.URLParameters(RABBITMQ_URL))
        channel = connection.channel()
        channel.queue_declare(queue=QUEUE_NAME, durable=True)
        return channel
    except pika.exceptions.AMQPError as e:
        logger.error(f"RabbitMQ connection error: {e}")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="RabbitMQ connection error",
        )


@asynccontextmanager
async def lifespan(app: FastAPI):
    """Initialize and clean up RabbitMQ resources"""
    app.state.rabbitmq_channel = setup_rabbitmq()
    yield

    if hasattr(app.state, "rabbitmq_channel"):
        app.state.rabbitmq_channel.close()
        logger.info("RabbitMQ connection closed")


app = FastAPI(lifespan=lifespan, title="RabbitMQ Producer API", version="1.0.0")


@app.get("/send")
async def send_message(msg: Optional[str] = None):
    """Publish a message to RabbitMQ queue"""
    if not msg:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Message cannot be empty",
        )
    try:
        app.state.rabbitmq_channel.basic_publish(
            exchange="",
            routing_key=QUEUE_NAME,
            body=msg.encode(),
            properties=pika.BasicProperties(delivery_mode=pika.DeliveryMode.Persistent),
        )
        logger.info(f"Message sent: {msg}")
        return JSONResponse(
            content={"message": msg, "status": "Message sent"},
            status_code=status.HTTP_200_OK,
        )
    except pika.exceptions.AMQPError as e:
        logger.error(f"Message publish failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Message publish failed",
        )


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
